.gitlab-ci.yml

```diff
@@ -187,6 +187,54 @@ spellcheck:
         cargo spellcheck check -vvv --cfg=.config/cargo_spellcheck.toml --checkers hunspell --code 1 examples/delegator/${contract}/;
       done
 
+codecov:
+  stage:                           workspace
+  <<:                              *docker-env
+  <<:                              *test-refs
+  needs:
+    - job:                         check-std
+      artifacts:                   false
+  variables:
+    # For codecov it's sufficient to run the fuzz tests only once.
+    QUICKCHECK_TESTS:              1
+    # Now we're using the MIR source-based Rust test coverage, the setup assumes that default
+    # toolchain is nightly and both `grcov` and `rustup component add llvm-tools-preview` are
+    # installed on it.
+    RUSTFLAGS:                     "-Zinstrument-coverage"
+    LLVM_PROFILE_FILE:             "llvmcoveragedata-%p-%m.profraw"
+    CARGO_INCREMENTAL:             0
+    # The `cargo-taurpalin` coverage reporting tool seems to have better code instrumentation and thus
+    # produces better results for Rust codebases in general. However, unlike `grcov` it requires
+    # running docker with `--security-opt seccomp=unconfined` which is why we use `grcov` instead.
+  before_script:
+    - *rust-info-script
+    # RUSTFLAGS are the cause target cache can't be used here
+    - unset "CARGO_TARGET_DIR"
+    - cargo clean
+  script:
+    - cargo build --verbose --all-features --workspace
+    - cargo test --verbose --all-features --no-fail-fast --workspace
+
+    # Just needed as long as we have the `ink-experimental-engine` feature.
+    # We must additionally run the coverage without `--all-features` here -- this
+    # would imply the feature `ink-experimental-engine`. So in order to still run
+    # the tests without the experimental engine feature we need this command.
+    - cargo test --verbose --features std --no-fail-fast --workspace
+
+    # coverage with branches
+    - grcov . --binary-path ./target/debug/ --source-dir . --output-type lcov --llvm --branch
+        --ignore-not-existing --ignore "/*" --ignore "tests/*" --output-path lcov-w-branch.info
+    - rust-covfix lcov-w-branch.info --output lcov-w-branch-fixed.info
+    # We'd like to not use a remote bash script for uploading the coverage reports,
+    # binary named `codecov` is already installed in the CI image, just doesn't work
+    # https://github.com/codecov/uploader/issues/217 https://github.com/codecov/uploader/issues/222
```

Comment:
Why doesn't it work on the CI image?

Reply:
because their binary is broken and I found those mentioned issues to prove it.

Reply:
Gotcha, I missed the mentioned issues, my bad. 

One nitpick here then: I'd mention something along the lines of "See issues below for the cause..." and have the issues in bullet points.

I noticed in https://github.com/codecov/uploader/issues/217 they made you open https://github.com/codecov/uploader/issues/228, so we should also update that.

We should also create an ink! CI issue so that we can keep follow up once these issues are closed.